Port CUDA torch.geqrf to ATen

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56257 Fix MAGMA qr for empty batched inputs
* #56256 Add cuSOLVER path for torch.linalg.qr
* #56255 Remove size arguments for internal orgqr and geqrf calls
* #56254 Add non-allocating helper function for torch.linalg.qr
* #56253 Add cuBLAS path for batched torch.geqrf
* #56252 Add cuSOLVER path for torch.geqrf
* **#56251 Port CUDA torch.geqrf to ATen**

This PR ports `torch.geqrf` from TH to ATen for CUDA path.

Resolves https://github.com/pytorch/pytorch/issues/24569

Differential Revision: [D27960155](https://our.internmc.facebook.com/intern/diff/D27960155)

IvanYashchuk committed:
Port CUDA torch.geqrf to ATen

This PR ports `torch.geqrf` from TH to ATen for CUDA path.

Resolves https://github.com/pytorch/pytorch/issues/24569

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port CUDA torch.geqrf to ATen"


This PR ports `torch.geqrf` from TH to ATen for CUDA path.

Resolves https://github.com/pytorch/pytorch/issues/24569

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5716,14 +5716,12 @@
   dispatch:
-    CPU: geqrf
-    CUDA: legacy::cuda::_th_geqrf
+    CPU, CUDA: geqrf
```
You don't have to do it this PR, but consider making this structured! https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md

ngimel commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1999,6 +1999,53 @@ Tensor& orgqr_kernel_impl(Tensor& result, const Tensor& tau, int64_t n_columns)
+
+  // magmaGeqrf uses a hybrid CPU-GPU algorithm to compute the elementary reflectors.
+  // The driver routine geqrf2_gpu accepts a tensor on the CPU for elementary reflectors.
+  Tensor tau_cpu = at::empty(tau.sizes(), tau.options().device(at::kCPU));
```
does it make sense to allocate this in pinned memory and use non-blocking copy below? The effect is probably negligible, but I don't know off the top of my head.

IvanYashchuk replied:
I think it wouldn't hurt.
Is it simply `Tensor tau_cpu = at::empty(tau.sizes(), tau.options().device(at::kCPU).pinned_memory(true));` ?

ngimel replied:
Yes, and `,/*non_blocking*/true` for `copy_` below, iirc.

IvanYashchuk replied:
Thank you! It did give non-negligible speedup:
```
|                              | before | after |
|------------------------------|--------|-------|
| torch.Size([2, 2])           | 3.0    | 2.3   |
| torch.Size([2, 2, 2])        | 5.8    | 4.4   |
| torch.Size([32, 2, 2])       | 88.3   | 70.6  |
| torch.Size([8, 8])           | 2.9    | 2.2   |
| torch.Size([2, 8, 8])        | 5.7    | 4.4   |
| torch.Size([32, 8, 8])       | 88.3   | 70.7  |
| torch.Size([16, 16])         | 3.0    | 2.2   |
| torch.Size([2, 16, 16])      | 5.7    | 4.5   |
| torch.Size([32, 16, 16])     | 89.0   | 72.4  |
| torch.Size([32, 32])         | 3.0    | 2.3   |
| torch.Size([2, 32, 32])      | 5.8    | 4.6   |
| torch.Size([32, 32, 32])     | 91.6   | 72.7  |
| torch.Size([64, 64])         | 3.1    | 2.4   |
| torch.Size([2, 64, 64])      | 6.0    | 4.7   |
| torch.Size([32, 64, 64])     | 93.1   | 74.0  |
| torch.Size([128, 128])       | 3.5    | 2.8   |
| torch.Size([2, 128, 128])    | 6.9    | 5.5   |
| torch.Size([32, 128, 128])   | 107.8  | 88.1  |
| torch.Size([256, 256])       | 4.7    | 3.7   |
| torch.Size([2, 256, 256])    | 9.2    | 7.2   |
| torch.Size([32, 256, 256])   | 144.1  | 112.1 |
| torch.Size([512, 512])       | 6.9    | 6.0   |
| torch.Size([2, 512, 512])    | 13.3   | 11.8  |
| torch.Size([32, 512, 512])   | 209.3  | 184.6 |
| torch.Size([1024, 1024])     | 15.4   | 14.6  |
| torch.Size([2, 1024, 1024])  | 30.5   | 28.9  |
| torch.Size([32, 1024, 1024]) | 486.5  | 462.4 |

Times are in milliseconds (ms).
```

ngimel replied:
I just hope magma is synchronizing when it is writing tau, otherwise nonblocking copy might produce wrong result (it will copy tau to gpu before it's properly populated)

IvanYashchuk committed:
Update on "Port CUDA torch.geqrf to ATen"


This PR ports `torch.geqrf` from TH to ATen for CUDA path.

Resolves https://github.com/pytorch/pytorch/issues/24569

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port CUDA torch.geqrf to ATen"


This PR ports `torch.geqrf` from TH to ATen for CUDA path.

Resolves https://github.com/pytorch/pytorch/issues/24569

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port CUDA torch.geqrf to ATen"


This PR ports `torch.geqrf` from TH to ATen for CUDA path.

Resolves https://github.com/pytorch/pytorch/issues/24569

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port CUDA torch.geqrf to ATen"


This PR ports `torch.geqrf` from TH to ATen for CUDA path.

Resolves https://github.com/pytorch/pytorch/issues/24569

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port CUDA torch.geqrf to ATen"


This PR ports `torch.geqrf` from TH to ATen for CUDA path.

Resolves https://github.com/pytorch/pytorch/issues/24569

[ghstack-poisoned]

IvanYashchuk commented:
@ngimel, @mruberry I updated this PR to use pinned memory and non-blocking copy for the `tau_cpu` tensor that is passed to MAGMA.

mruberry commented:
Hey @IvanYashchuk, unfortunately the internal tools are unable to merge this; would you try rebasing the stack now that the first 2 prs are in?

IvanYashchuk committed:
Update on "Port CUDA torch.geqrf to ATen"


This PR ports `torch.geqrf` from TH to ATen for CUDA path.

Resolves https://github.com/pytorch/pytorch/issues/24569

Differential Revision: [D27960155](https://our.internmc.facebook.com/intern/diff/D27960155)

[ghstack-poisoned]

IvanYashchuk commented:
@mruberry, I rebased the stack.
